modules/python3gl/src/volumepytoglconverter.cpp

```diff
@@ -0,0 +1,116 @@
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ *********************************************************************************/
+
+#include <inviwo/python3gl/volumepytoglconverter.h>
+
+namespace inviwo {
+
+std::shared_ptr<VolumeGL> VolumePy2GLConverter::createFrom(
+    std::shared_ptr<const VolumePy> volumeSrc) const {
+    auto volumeDst = std::make_shared<VolumeGL>(
+        volumeSrc->getDimensions(), volumeSrc->getDataFormat(), volumeSrc->getSwizzleMask(),
+        volumeSrc->getInterpolation(), volumeSrc->getWrapping(), false);
+
+    const auto tex = volumeDst->getTexture();
+    auto srcSize = static_cast<size_t>(volumeSrc->data().nbytes());
+    auto dstSize = tex->getNumberOfValues() * tex->getSizeInBytes();
+
+    if (srcSize == dstSize &&
+        pybind11::array::c_style == (volumeSrc->data().flags() & pybind11::array::c_style)) {
+        auto src = volumeSrc->data().data(0);
+        volumeDst->getTexture()->initialize(src);
+    } else {
+        throw Exception("Unable to convert from VolumePy to VolumeGL", IVW_CONTEXT);
+    }
+
+    return volumeDst;
+}
+
+void VolumePy2GLConverter::update(std::shared_ptr<const VolumePy> volumeSrc,
+                                  std::shared_ptr<VolumeGL> volumeDst) const {
+    volumeDst->setDimensions(volumeSrc->getDimensions());
+    volumeDst->setSwizzleMask(volumeSrc->getSwizzleMask());
+    volumeDst->setInterpolation(volumeSrc->getInterpolation());
+    volumeDst->setWrapping(volumeSrc->getWrapping());
+
+    const auto tex = volumeDst->getTexture();
+    auto srcSize = static_cast<size_t>(volumeSrc->data().nbytes());
+    auto dstSize = tex->getNumberOfValues() * tex->getSizeInBytes();
+
+    if (srcSize == dstSize &&
+        pybind11::array::c_style == (volumeSrc->data().flags() & pybind11::array::c_style)) {
+        auto src = volumeSrc->data().data(0);
+        volumeDst->getTexture()->upload(src);
+    } else {
+        throw Exception("Unable to convert from VolumePy to VolumeGL", IVW_CONTEXT);
+    }
+}
+
+std::shared_ptr<VolumePy> VolumeGL2PyConverter::createFrom(
+    std::shared_ptr<const VolumeGL> volumeSrc) const {
+    auto volumeDst = std::make_shared<VolumePy>(
+        volumeSrc->getDimensions(), volumeSrc->getDataFormat(), volumeSrc->getSwizzleMask(),
+        volumeSrc->getInterpolation(), volumeSrc->getWrapping());
+
+    const auto tex = volumeSrc->getTexture();
```

Comment:
is it guaranteed that the correct context is active?

Reply:
No we don't guarantee that